src/ngx_http_lua_fd.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (C) 2014 Daurnimator
+ */
+
+/* need to include DDEBUG so that ngx_http_lua_util.h works */
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+#include "ngx_http_lua_fd.h"
+
+#include <math.h> /* HUGE_VAL */
+#include <stdlib.h> /* calloc, free */
+#include <poll.h> /* POLLIN, POLLOUT */
+
+#include <lua.h>
+#include <lauxlib.h>
+
+#include "ngx_core.h"
+#include "ngx_http.h" /* ngx_http_request_t, ngx_http_run_posted_requests */
+#include "ngx_http_lua_common.h" /* ngx_http_lua_module, ngx_http_lua_co_ctx_t */
+
+#include "ngx_http_lua_util.h" /* ngx_http_lua_get_lua_vm, ngx_http_lua_run_thread, ngx_http_lua_finalize_request */
+#include "ngx_http_lua_contentby.h" /* ngx_http_lua_content_wev_handler */
+
+typedef struct {
+    ngx_http_request_t *request;
+    ngx_http_lua_co_ctx_t *co_ctx;
+    ngx_connection_t *conn;
+} ngx_http_lua_udata_t;
+
+static ngx_int_t ngx_http_lua_fd_resume_request(ngx_http_request_t *r) {
+    ngx_http_lua_ctx_t *ctx;
+    lua_State *vm;
+    ngx_int_t rc;
+    if ((ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module)) == NULL) {
+        return NGX_ERROR;
+    }
+    /* restore normal resume handler */
+    ctx->resume_handler = ngx_http_lua_wev_handler;
+    /* resume lua thread */
+    vm = ngx_http_lua_get_lua_vm(r, ctx);
+    rc = ngx_http_lua_run_thread(vm, r, ctx, 0 /*nret*/);
+    switch (rc) {
+        case NGX_DONE: /* coroutine finished */
+            ngx_http_lua_finalize_request(r, NGX_DONE);
+            /* fall-through */
+        case NGX_AGAIN: /* coroutine yielded */
+            return ngx_http_lua_run_posted_threads(r->connection, vm, r, ctx);
+        default: /* NGX_ERROR: coroutine failed */
+            if (ctx->entered_content_phase) {
+                ngx_http_lua_finalize_request(r, rc);
+                return NGX_DONE;
+            }
+            return rc;
+    }
+}
+
+static void ngx_http_lua_fd_rev_handler(ngx_event_t *ev) {
+    ngx_connection_t *conn = ev->data;
+    ngx_http_lua_udata_t *u = conn->data;
+    ngx_http_request_t *r = u->request;
+    ngx_http_lua_co_ctx_t *co_ctx = u->co_ctx;
+    ngx_http_lua_ctx_t *ctx;
+
+
+    ngx_http_lua_cleanup_pending_operation(co_ctx);
+    ev = NULL, conn = NULL, u = NULL; /* now invalidated */
+
+    if ((ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module)) != NULL) {
```

Comment:
Style: better avoid using the side-effect of C assignments according to nginx's coding style.